### PR TITLE
xftp: send file descriptions when ready

### PR DESCRIPTION
**Pull request by epoberezkin**

### CODE REVIEW

File: src/Simplex/Chat.hs [hunk 1]
```diff
@@ -2142,51 +2154,53 @@ processAgentMsgSndFile _corrId aFileId msg =
...
-                -- update sent file status
-                -- ??? possibly another event as we need one event per group, not per member
-                -- toView $ CRSndFileComplete user ci ft
-                pure ()
-              _ -> pure () -- TODO error
-            _ -> pure () -- TODO error
-          pure ()
+            (Just sharedMsgId, Nothing) -> do
+              (ft, sfts) <- withStore $ \db -> getSndFileTransfer db user fileId
+              when (length sfts < length rfds) $ throwChatError $ CEInternalError "not enough XFTP file descriptions to send"
```
Comment (spaced4ndy) [inline src/Simplex/Chat.hs:2169]: shouldn't be like this?
```suggestion
              when (length rfds < length sfts) $ throwChatError $ CEInternalError "not enough XFTP file descriptions to send"
```

File: src/Simplex/Chat.hs [hunk 2]
```diff
@@ -527,10 +529,20 @@ processChatCommand = \case
...
         let ciFile = CIFile {fileId, fileName, fileSize, filePath = Just file, fileStatus = CIFSSndStored}
+        case contactOrGroup of
+          CGContact Contact {activeConn} -> withStore' $ \db -> createSndFTDescrXFTP db user Nothing activeConn ft fileDescr
+          CGGroup (Group _ ms) -> forM_ ms $ \m -> sendToMember m `catchError` (toView . CRChatError (Just user))
+            where
+              -- we are not sending files to pending members, same as with inline files
+              sendToMember m@GroupMember {activeConn = Just conn@Connection {connStatus}} =
+                when ((connStatus == ConnReady || connStatus == ConnSndReady) && not (connDisabled conn)) $
+                  withStore' $ \db -> createSndFTDescrXFTP db user (Just m) conn ft fileDescr
+              sendToMember _ = pure ()
```
Comment (spaced4ndy) [inline src/Simplex/Chat.hs:545]: this doesn't seem to send anything
```suggestion
          CGGroup (Group _ ms) -> forM_ ms $ \m -> saveMemberFD m `catchError` (toView . CRChatError (Just user))
            where
              -- we are not sending files to pending members, same as with inline files
              saveMemberFD m@GroupMember {activeConn = Just conn@Connection {connStatus}} =
                when ((connStatus == ConnReady || connStatus == ConnSndReady) && not (connDisabled conn)) $
                  withStore' $ \db -> createSndFTDescrXFTP db user (Just m) conn ft fileDescr
              saveMemberFD _ = pure ()
```

File: src/Simplex/Chat/Store.hs [hunk 3]
```diff
@@ -2740,22 +2743,43 @@ createSndFileTransferXFTP db User {userId} contactOrGroup filePath FileInvitatio
   fileId <- insertedRowId db
   pure FileTransferMeta {fileId, xftpSndFile, fileName, filePath, fileSize, fileInline = Nothing, chunkSize, cancelled = False}
 
-createSndDirectFTDescrXFTP :: DB.Connection -> User -> Contact -> FileTransferMeta -> Text -> IO ()
-createSndDirectFTDescrXFTP db User {userId} Contact {activeConn = Connection {connId}} FileTransferMeta {fileId} rfdText = do
-  let fileStatus = FSConnected
-  DB.execute db "INSERT INTO xftp_file_descriptions (user_id, file_descr_text, file_descr_complete) VALUES (?,?,?)" (userId, rfdText, True)
+createSndFTDescrXFTP :: DB.Connection -> User -> Maybe GroupMember -> Connection -> FileTransferMeta -> FileDescr -> IO ()
+createSndFTDescrXFTP db User {userId} m Connection {connId} FileTransferMeta {fileId} FileDescr {fileDescrText, fileDescrPartNo, fileDescrComplete} = do
```
Comment (spaced4ndy) [inline src/Simplex/Chat/Store.hs:2747]:
```suggestion
createSndFD :: DB.Connection -> User -> Maybe GroupMember -> Connection -> FileTransferMeta -> FileDescr -> IO ()
createSndFD db User {userId} m Connection {connId} FileTransferMeta {fileId} FileDescr {fileDescrText, fileDescrPartNo, fileDescrComplete} = do
```

File: src/Simplex/Chat/Store.hs [hunk 4]
```diff
@@ -2740,22 +2743,43 @@ createSndFileTransferXFTP db User {userId} contactOrGroup filePath FileInvitatio
...
   fileDescrId <- insertedRowId db
   DB.execute
     db
-    "INSERT INTO snd_files (file_id, file_status, file_descr_id, connection_id) VALUES (?,?,?,?)"
-    (fileId, fileStatus, fileDescrId, connId)
+    "INSERT INTO snd_files (file_id, file_status, file_descr_id, group_member_id, connection_id) VALUES (?,?,?,?,?)"
+    (fileId, fileStatus, fileDescrId, groupMemberId' <$> m, connId)
+
+updateSndFTDescrXFTP :: DB.Connection -> User -> SndFileTransfer -> Text -> IO ()
+updateSndFTDescrXFTP db user@User {userId} sft@SndFileTransfer {fileId, fileDescrId} rfdText = do
```
Comment (spaced4ndy) [inline src/Simplex/Chat/Store.hs:2760]:
```suggestion
updateSndFD :: DB.Connection -> User -> SndFileTransfer -> Text -> IO ()
updateSndFD db user@User {userId} sft@SndFileTransfer {fileId, fileDescrId} rfdText = do
```

File: src/Simplex/Chat/Store.hs (hunk 3 above)
Comment (spaced4ndy) [reply src/Simplex/Chat/Store.hs:2747]: not necessarily better, just follows naming of createRcvFD_

File: src/Simplex/Chat/Store.hs [hunk 5]
```diff
@@ -2740,22 +2743,43 @@ createSndFileTransferXFTP db User {userId} contactOrGroup filePath FileInvitatio
...
+  DB.execute
+    db
+    [sql|
+      UPDATE xftp_file_descriptions
+      SET file_descr_text = ?, file_descr_part_no = ?, file_descr_complete = ?
+      WHERE user_id = ? AND file_descr_id = ?
+    |]
+    (rfdText, 1 :: Int, True, userId, fileDescrId)
```
[the remark this answers is not in the document]
Comment (epoberezkin) [reply src/Simplex/Chat/Store.hs:2768]: they are still recipient's descriptions and the descriptions you received can be forwarded, so it's the same entity...

**Pull request merged**